client/components/happychat/timeline.jsx

```diff
@@ -0,0 +1,169 @@
+ * Otherwise just return a single paragraph with the text.
+ */
+const messageText = when( linksNotEmpty, messageWithLinks, messageParagraph );
+const messageAvatar = when( propExists( 'meta.image' ), ( { meta } ) => <img alt={ meta.nick } src={ meta.image } /> );
+
+/*
+ * Group messages based on user so when any user sends multiple messages they will be grouped
+ * within the same message bubble until it reaches a message from a different user.
+ */
+const renderGroupedMessages = ( { item, isCurrentUser }, index ) => {
+	const [ event, ... rest ] = item;
+	const userAvatar = messageAvatar( { meta: event } );
+	return (
+		<div className={ classnames( 'happychat__timeline-message', { 'user-message': isCurrentUser } ) } key={ event.id || index }>
+			<div className="happychat__message-text">
+				{ messageText( {
+					message: event.message,
+					nick: event.nick,
+					key: event.id,
+					links: event.links
+				} ) }
+				{ rest.map( ( { message, id: key, links } ) => messageText( { message, key, links } ) ) }
+			</div>
+			<div className="happychat__message-meta">
+				<div className="happychat__message-avatar">
+				{ isCurrentUser ? userAvatar : <AgentW /> }
+				</div>
+			</div>
+		</div>
+	);
+};
+
+const itemTypeIs = type => ( { item: [ firstItem ] } ) => firstItem.type === type;
+
+/*
+ * Renders a chat bubble with multiple messages grouped by user.
+ */
+const renderGroupedTimelineItem = first(
+	when( itemTypeIs( 'message' ), renderGroupedMessages ),
+	( { item } ) => debug( 'no handler for message type', item[ 0 ][ 1 ].type, item )
+);
+
+const groupMessages = messages => {
+	const grouped = messages.reduce( ( { user_id, type, group, groups }, message ) => {
+		const message_user_id = message.user_id;
+		const message_type = message.type;
+		if ( user_id !== message_user_id || message_type !== type ) {
+			return {
+				user_id: message_user_id,
+				type: message_type,
+				group: [ message ],
+				groups: group ? groups.concat( [ group ] ) : groups
+			};
+		}
+		// it's the same user so group it together
+		return { user_id, group: group.concat( [ message ] ), groups, type };
+	}, { groups: [] } );
+
+	return grouped.groups.concat( [ grouped.group ] );
+};
+
+const welcomeMessage = () => (
+	<div className="happychat__welcome">
+		{ translate( 'This is the beginning of your chat history with WordPress.com support. A chat history will be stored here.' ) }
+	</div>
+);
+
+const timelineHasContent = ( { timeline } ) => isArray( timeline ) && ! isEmpty( timeline );
+
+const renderTimeline = ( { timeline, isCurrentUser, onScrollContainer, scrollbleedLock, scrollbleedUnlock } ) => (
+	<div className="happychat__conversation"
+		ref={ onScrollContainer }
+		onMouseEnter={ scrollbleedLock }
+		onMouseLeave={ scrollbleedUnlock }>
+		{ groupMessages( timeline ).map( item => renderGroupedTimelineItem( {
+			item,
+			isCurrentUser: isCurrentUser( item[ 0 ] )
```

Comment:
Actually, the `groupMessages` turns an array of messages into multiple arrays segmented by user id. So in this case it's still an array.